test/ha/activator_test.go

```diff
@@ -0,0 +1,146 @@
+// that we can scale from zero after activator restart.
+func TestActivatorHA(t *testing.T) {
+	clients := e2e.Setup(t)
+
+	pods, err := clients.KubeClient.Kube.CoreV1().Pods(servingNamespace).List(metav1.ListOptions{
+		LabelSelector: activatorLabel,
+	})
+	if err != nil {
+		t.Fatalf("Failed to get activator pods: %v", err)
+	}
+	// make sure we have the right number of activator replicas before continuing
+	if len(pods.Items) != haReplicas {
+		t.Skip(fmt.Sprintf("The test requires %d activator pods, got: %d", haReplicas, len(pods.Items)))
+	}
+	// sort the pods according to creation timestamp so that we can always kill the oldest one
+	sort.Slice(pods.Items, func(i, j int) bool { return pods.Items[i].CreationTimestamp.Before(&pods.Items[j].CreationTimestamp) })
+
+	// create first service that we will continually probe during activator restart
+	names, resources := createPizzaPlanetService(t,
+		rtesting.WithConfigAnnotations(map[string]string{
+			autoscaling.MinScaleAnnotationKey:  "1",  //make sure we don't scale to zero during the test
+			autoscaling.TargetBurstCapacityKey: "-1", // make sure all requests go through the activator
+		}),
+	)
+	test.CleanupOnInterrupt(func() { test.TearDown(clients, names) })
+	defer test.TearDown(clients, names)
+
+	// create second service that will be scaled to zero and after stopping the activator we'll
+	// ensure it can be scaled back from zero
+	namesScaleToZero, resourcesScaleToZero := createPizzaPlanetService(t,
+		rtesting.WithConfigAnnotations(map[string]string{
+			autoscaling.WindowAnnotationKey:    autoscaling.WindowMin.String(), //make sure we scale to zero quickly
+			autoscaling.TargetBurstCapacityKey: "-1",                           // make sure all requests go through the activator
+		}),
+	)
+	test.CleanupOnInterrupt(func() { test.TearDown(clients, namesScaleToZero) })
+	defer test.TearDown(clients, namesScaleToZero)
+
+	if err := e2e.WaitForScaleToZero(t, revisionresourcenames.Deployment(resourcesScaleToZero.Revision), clients); err != nil {
+		t.Fatalf("Failed to scale to zero: %v", err)
+	}
+
+	scaleToZeroURL := resources.Service.Status.URL.URL()
+	prober := test.RunRouteProber(log.Printf, clients, resources.Service.Status.URL.URL())
+	defer test.AssertProberDefault(t, prober)
+
+	spoofingClient, err := pkgTest.NewSpoofingClient(clients.KubeClient, t.Logf, scaleToZeroURL.Hostname(), test.ServingFlags.ResolvableDomain, test.AddRootCAtoTransport(t.Logf, clients, test.ServingFlags.Https))
+	if err != nil {
+		t.Fatalf("Error creating spoofing client: %v", err)
+	}
+
+	activatorPod := pods.Items[0].Name // stop the oldest activator pod
+
+	var gracePeriodSeconds int64 = 0
+	clients.KubeClient.Kube.CoreV1().Pods(servingNamespace).Delete(activatorPod, &metav1.DeleteOptions{
+		GracePeriodSeconds: &gracePeriodSeconds,
+	})
+
+	if err := waitForPublicEndpointAddresses(t, clients, resourcesScaleToZero.Revision.Name,
```

Comment:
What are we waiting here for? Can you add a comment?
Since won't the k8s HPA immediately start another pod?

Reply:
We wait until the Knative service under test doesn't have the killed activator among it's endpoints. The check happens often (100ms period) so it should not hit the point when the old activator is gone and the newly started is activator is in the list immediately. There should be a little gap between the moment when the old one is removed and the new one is added. Or no?

Reply:
I've just verified that the interval between the service having 1 endpoints and 2 endpoints back again is about 12 seconds (on my laptop using Minikube):
```
Got 1 endpoints at 2020-04-07 08:40:45.304490524 +0200 CEST m=+49.323762516
Got 2 endpoints at 2020-04-07 08:40:57.933583602 +0200 CEST m=+61.952855599
```
So, this function should be safe.

Reply:
Actually, I think I hit the problem you were mentioning in the last CI run. So added one more commit to not rely on the number of endpoints.